src/testing/starknet_state.rs

```diff
@@ -0,0 +1,449 @@
+use super::{starknet_state_error::StarknetStateError, type_utils::ExecutionInfo};
+use std::collections::HashMap;
+
+use felt::Felt;
+use num_traits::Zero;
+
+use crate::{
+    business_logic::{
+        execution::{
+            execution_entry_point::ExecutionEntryPoint,
+            objects::{CallInfo, Event, TransactionExecutionContext, TransactionExecutionInfo},
+        },
+        fact_state::{
+            in_memory_state_reader::InMemoryStateReader, state::ExecutionResourcesManager,
+        },
+        state::{
+            cached_state::CachedState,
+            state_api::{State, StateReader},
+        },
+        transaction::{
+            error::TransactionError,
+            objects::{
+                internal_declare::InternalDeclare, internal_deploy::InternalDeploy,
+                internal_invoke_function::InternalInvokeFunction,
+            },
+            transactions::Transaction,
+        },
+    },
+    definitions::{constants::TRANSACTION_VERSION, general_config::StarknetGeneralConfig},
+    services::api::{
+        contract_class::{ContractClass, EntryPointType},
+        messages::StarknetMessageToL1,
+    },
+    utils::Address,
+};
+
+// ---------------------------------------------------------------------
+/// StarkNet testing object. Represents a state of a StarkNet network.
+pub(crate) struct StarknetState {
+    pub(crate) state: CachedState<InMemoryStateReader>,
+    pub(crate) general_config: StarknetGeneralConfig,
+    l2_to_l1_messages: HashMap<Vec<u8>, usize>,
+    l2_to_l1_messages_log: Vec<StarknetMessageToL1>,
+    events: Vec<Event>,
+}
+
+impl StarknetState {
+    #![allow(unused)] // TODO: delete once used
+    pub fn new(config: Option<StarknetGeneralConfig>) -> Self {
+        let general_config = config.unwrap_or_default();
+        let state_reader = InMemoryStateReader::new(HashMap::new(), HashMap::new());
+
+        let state = CachedState::new(state_reader, Some(HashMap::new()));
+
+        let l2_to_l1_messages = HashMap::new();
+        let l2_to_l1_messages_log = Vec::new();
+
+        let events = Vec::new();
+        StarknetState {
+            state,
+            general_config,
+            l2_to_l1_messages,
+            l2_to_l1_messages_log,
+            events,
+        }
+    }
+
+    // ------------------------------------------------------------------------------------
+    /// Declares a contract class.
+    /// Returns the class hash and the execution info.
+    /// Args:
+    /// contract_class - a compiled StarkNet contract
+    pub fn declare(
+        &mut self,
+        contract_class: ContractClass,
+    ) -> Result<([u8; 32], TransactionExecutionInfo), TransactionError> {
+        let tx = InternalDeclare::new(
+            contract_class.clone(),
+            self.chain_id(),
+            Address(1.into()),
```

Comment:
I don't think this line should be correct, since we're hardcoding an address to `1` in non-testing code.

However, if it is, then:
```suggestion
            Address(Felt::one()),
```

Reply:
The behavior es replicated from python code, they use a default sender address to create the create the declare, i am unsure why regardless.